src/plugins/ui_actions/public/service/ui_actions_service.ts

```diff
@@ -76,20 +91,68 @@ export class UiActionsService {
     return trigger.contract;
   };
 
-  public readonly registerAction = <T extends ActionType>(action: ActionByType<T>) => {
-    if (this.actions.has(action.id)) {
-      throw new Error(`Action [action.id = ${action.id}] already registered.`);
+  public readonly registerAction = <A extends AnyActionDefinition>(
+    definition: A
+  ): ActionInternal<A> => {
+    if (this.actions.has(definition.id)) {
+      throw new Error(`Action [action.id = ${definition.id}] already registered.`);
     }
 
+    const action = new ActionInternal(definition);
+
     this.actions.set(action.id, action);
+
+    return action;
+  };
+
+  protected readonly unregisterAction = (actionId: string): void => {
+    if (!this.actions.has(actionId)) {
+      throw new Error(`Action [action.id = ${actionId}] is not registered.`);
+    }
+
+    this.actions.delete(actionId);
+  };
+
+  public readonly addTriggerAction = <TriggerId extends keyof TriggerContextMapping>(
+    triggerId: TriggerId,
+    definition: ActionDefinition<TriggerContextMapping[TriggerId]>
+  ) => {
+    // Check if trigger exists, if not, next line throws.
+    this.getTrigger(triggerId);
+
+    const action = this.registerAction(definition);
+    this.__attachAction(triggerId, action.id);
+
+    return action;
+  };
+
+  public readonly removeTriggerAction = <TriggerId extends keyof TriggerContextMapping>(
+    triggerId: TriggerId,
+    actionId: string
+  ) => {
+    this.detachAction(triggerId, actionId);
+    this.unregisterAction(actionId);
+  };
+
+  // public readonly removeTriggerAction =
+
+  protected readonly __attachAction = <TriggerId extends keyof TriggerContextMapping>(
+    triggerId: TriggerId,
+    actionId: string
+  ): void => {
+    const actionIds = this.triggerToActions.get(triggerId);
+
+    if (!actionIds!.find(id => id === actionId)) {
+      this.triggerToActions.set(triggerId, [...actionIds!, actionId]);
+    }
   };
 
-  public readonly getAction = <T extends ActionType>(id: string): ActionByType<T> => {
+  public readonly getAction = <T extends AnyActionDefinition>(id: string): ActionInternal<T> => {
```

Comment:
Why did this change to ActionInternal instead of ActionByType?  Shouldn't internal be internal?

Reply:
I've changed it to Action. The idea was to improve `Action` interface that's why I created `ActionInternal`, I would have used `Action`, but it is already taken.